// just-task/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.clearCache = exports.argv = exports.option = exports.resolveCwd = exports.resolve = exports.resetResolvePaths = exports.addResolvePath = exports.condition = void 0;
__exportStar(require("./undertaker"), exports);
__exportStar(require("./task"), exports);
__exportStar(require("./interfaces"), exports);
var condition_1 = require("./condition");
Object.defineProperty(exports, "condition", { enumerable: true, get: function () { return condition_1.condition; } });
var resolve_1 = require("./resolve");
Object.defineProperty(exports, "addResolvePath", { enumerable: true, get: function () { return resolve_1.addResolvePath; } });
Object.defineProperty(exports, "resetResolvePaths", { enumerable: true, get: function () { return resolve_1.resetResolvePaths; } });
Object.defineProperty(exports, "resolve", { enumerable: true, get: function () { return resolve_1.resolve; } });
Object.defineProperty(exports, "resolveCwd", { enumerable: true, get: function () { return resolve_1.resolveCwd; } });
var option_1 = require("./option");
Object.defineProperty(exports, "option", { enumerable: true, get: function () { return option_1.option; } });
Object.defineProperty(exports, "argv", { enumerable: true, get: function () { return option_1.argv; } });
var cache_1 = require("./cache");
Object.defineProperty(exports, "clearCache", { enumerable: true, get: function () { return cache_1.clearCache; } });
__exportStar(require("./logger"), exports);
__exportStar(require("./chain"), exports);
__exportStar(require("./watch"), exports);
