// just-scripts/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.spawn = exports.copyInstructions = exports.webpackMerge = exports.createStylesOverlay = exports.displayBailoutOverlay = exports.fileOverlay = exports.stylesOverlay = exports.htmlOverlay = exports.tsOverlay = exports.webpackOverlays = exports.taskPresets = void 0;
__exportStar(require("./tasks"), exports);
const taskPresets = require("./task-presets");
exports.taskPresets = taskPresets;
__exportStar(require("./webpack/webpack.config"), exports);
__exportStar(require("./webpack/webpack.serve.config"), exports);
// Webpack configs and overlays
const tsOverlay_1 = require("./webpack/overlays/tsOverlay");
Object.defineProperty(exports, "tsOverlay", { enumerable: true, get: function () { return tsOverlay_1.tsOverlay; } });
const htmlOverlay_1 = require("./webpack/overlays/htmlOverlay");
Object.defineProperty(exports, "htmlOverlay", { enumerable: true, get: function () { return htmlOverlay_1.htmlOverlay; } });
const stylesOverlay_1 = require("./webpack/overlays/stylesOverlay");
Object.defineProperty(exports, "stylesOverlay", { enumerable: true, get: function () { return stylesOverlay_1.stylesOverlay; } });
Object.defineProperty(exports, "createStylesOverlay", { enumerable: true, get: function () { return stylesOverlay_1.createStylesOverlay; } });
const fileOverlay_1 = require("./webpack/overlays/fileOverlay");
Object.defineProperty(exports, "fileOverlay", { enumerable: true, get: function () { return fileOverlay_1.fileOverlay; } });
const displayBailoutOverlay_1 = require("./webpack/overlays/displayBailoutOverlay");
Object.defineProperty(exports, "displayBailoutOverlay", { enumerable: true, get: function () { return displayBailoutOverlay_1.displayBailoutOverlay; } });
exports.webpackOverlays = {
    typescript: tsOverlay_1.tsOverlay,
    html: htmlOverlay_1.htmlOverlay,
    styles: stylesOverlay_1.stylesOverlay,
    file: fileOverlay_1.fileOverlay,
    displayBailout: displayBailoutOverlay_1.displayBailoutOverlay,
};
const webpackMerge = require("webpack-merge");
exports.webpackMerge = webpackMerge;
const copyInstructions = require("./copy/CopyInstruction");
exports.copyInstructions = copyInstructions;
__exportStar(require("just-task"), exports);
var just_scripts_utils_1 = require("just-scripts-utils");
Object.defineProperty(exports, "spawn", { enumerable: true, get: function () { return just_scripts_utils_1.spawn; } });
